src/ucs/datastruct/callbackq.h

```diff
@@ -0,0 +1,185 @@
+ */
+ucs_status_t ucs_callbackq_add_sync(ucs_callbackq_t *cbq, ucs_callback_t cb,
+                                    void *arg);
+
+
+/**
+ * Remove a callback from the queue immediately.
+ * This is *not* safe to call while another thread might be dispatching callbacks.
+ * However, it can be used from the dispatch context (e.g a callback may use this
+ * function to remove itself or another callback).
+ *
+ * If the pair (cb, arg) has a reference count > 1, the reference count is
+ * decremented by 1, and the callback is not removed.
+ *
+ * @param  [in] cbq      Callback queue to remove the callback from.
+ * @param  [in] cb       Callback to remove.
+ * @param  [in] arg      User-defined argument for the callback.
+ *
+ * @return UCS_ERR_NO_ELEM if element does not exist.
+ */
+ucs_status_t ucs_callbackq_remove_sync(ucs_callbackq_t *cbq, ucs_callback_t cb,
+                                       void *arg);
+
+
+/**
+ * Same as @ref ucs_callbackq_remove_sync, except it removes the callback even
+ * if its reference count is > 1, and guarantees the next time the callback
+ * queue is dispatched (on this thread), this callback will not be called, assuming
+ * it is called after @ref ucs_callbackq_add_async.
+ *
+ * @param  [in] cbq      Callback queue to remove the callback from.
+ * @param  [in] cb       Callback to remove.
+ * @param  [in] arg      User-defined argument for the callback.
+ */
+void ucs_callbackq_remove_all(ucs_callbackq_t *cbq, ucs_callback_t cb, void *arg);
+
+
+/**
+ * Add a callback to the queue.
+ * This can be used from any context and any thread, including but not limited to:
+ * - A callback can add another callback.
+ * - A thread can add a callback while another thread is dispatching callbacks.
+ *
+ * If the pair (cb, arg) already exists, it is not added, but its reference count
+ * is incremented.
+ *
+ * @param  [in] cbq      Callback queue to add the callback to.
+ * @param  [in] cb       Callback to add.
+ * @param  [in] arg      User-defined argument for the callback.
+ */
+ucs_status_t ucs_callbackq_add_async(ucs_callbackq_t *cbq, ucs_callback_t cb,
+                                     void *arg);
+
+
+/**
+ * Remove a callback from the queue in a lazy fashion. The callback will be
+ * removed at some point in the near future.
+ * This can be used from any context and any thread, including but not limited to:
+ * - A callback can remove another callback or itself.
+ * - A thread remove add a callback while another thread is dispatching callbacks.
+ *
+ * If the pair (cb, arg) has a reference count > 1, the reference count is
+ * decremented by 1, and the callback is not removed.
+ *
+ * @param  [in] cbq      Callback queue to remove the callback from.
+ * @param  [in] cb       Callback to remove.
+ * @param  [in] arg      User-defined argument for the callback.
+ */
+ucs_status_t ucs_callbackq_remove_async(ucs_callbackq_t *cbq, ucs_callback_t cb,
+                                        void *arg);
+
```

Comment:
imho it is better to use something else instead of xxx_sync and xxx_async because  sync and async already have different meaning in am callbacks